syntax = "proto3";

service EdgeStorage {
    rpc StoreFeatureMap (StoreFeatureMapRequest) returns (StoreFeatureMapReply) {}
    rpc FetchFeatureMap (FetchFeatureMapRequest) returns (FetchFeatureMapReply) {}
    rpc DeleteFeatureMap (DeleteFeatureMapRequest) returns (DeleteFeatureMapReply) {}
}

service NetworkSplit {
    rpc CloudCompute (CloudComputeRequest) returns (CloudComputeReply) {}
}

service EdgeRegister {
    rpc Register (RegisterRequest) returns (RegisterReply) {}
    rpc DeviceQuery (DeviceQueryRequest) returns (DeviceQueryReply) {}
}

message DeviceQueryRequest {}

message DeviceQueryReply {
    map<int32, string> device_addr = 1;
}

message RegisterRequest {
    string addr = 1;
}

message RegisterReply {
    int32 device_index = 1;
}

message StoreFeatureMapRequest {
    int32 track_id = 1;
    repeated double features = 2;
}

message StoreFeatureMapReply {
    bool success = 1;
}

message FetchFeatureMapRequest {
    int32 track_id = 1;
}

message FetchFeatureMapReply {
    bool success = 1;
    repeated double features = 2;
}

message DeleteFeatureMapRequest {
    int32 track_id = 1;
}

message DeleteFeatureMapReply {
    bool success = 1;
}

message CloudComputeRequest {
    int32 track_id = 1;
}

message CloudComputeReply {
    int32 label = 1;
}
